src/main/java/com/zegoggles/smssync/mail/MessageConverter.java

```diff
@@ -106,7 +107,28 @@ public MessageConverter(Context context, Preferences preferences,
         }
         final ConversionResult result = new ConversionResult(dataType);
         if (m != null) {
-            m.setFlag(Flag.SEEN, mMarkAsRead);
+            switch (mMarkAsReadType) {
+                case MESSAGE_STATUS:
+                    switch (dataType) {
+                        case SMS:
+                            m.setFlag(Flag.SEEN, "1".equals(msgMap.get(SmsConsts.READ)));
+                            break;
+                        case MMS:
+                            m.setFlag(Flag.SEEN, "1".equals(msgMap.get(MmsConsts.READ)));
+                            break;
+                        default:
+                            m.setFlag(Flag.SEEN, true);
+                            break;
+                    }
+                    break;
+                case UNREAD:
+                    m.setFlag(Flag.SEEN, false);
+                    break;
+                case READ:
+                default:
+                    m.setFlag(Flag.SEEN, true);
+                    break;
+            }
```

Comment:
drop all the duplicated `m.setFlag(...)` calls and extract to method with one single `m.setFlag(Flag.SEEN, markAsSeen(dataType, msgMap)` or similar